### PR TITLE
Symbol provider for bazel rules

### PR DESCRIPTION
**Pull request by damienpontifex**
Provide a [Go to symbol](https://code.visualstudio.com/docs/editor/editingevolved#_go-to-symbol) capability for navigating rules within a BUILD file

### CODE REVIEW

**Comment by damienpontifex** [comment]
I suppose the main thing I was looking for was a quick way to jump to a rule in a BUILD file. Loosely using it as file navigation vs 'symbols' in their pure sense. I wasn't aware of another way VSCode allowed that filter and jump apart from this

I like the link provider, especially if you could jump to a dependent rule. Would make navigating between bazel files much quicker

**Comment by allevato** [comment]
I've given this some more thought and I think we can bridge the two concepts without causing too much future pain after all.

If we create a symbol provider, we'll get some features like the "Outline" pane as part of that. That could be seen as a little redundant since there's also the "Bazel Targets Tree", but for people who use the "Outline" view regularly, it might provide good symmetry for just the targets in the open BUILD file. Likewise, the utility of the Command+Shift+O picker that you describe is good; the tree view can't replicate that.

Most BUILD files won't have much in the way of symbols aside from build targets—people shouldn't be defining functions there (they may define constants, and it would be nice to show those, but until we have a real language server, we'll have to skip those). Similarly, .bzl files won't have any targets, so we don't have to worry about it there.

But more importantly, representing targets as symbols may let us take advantage of other interesting VS Code UIs. For example, we could leverage "Find All References" on a target to render an `rdeps` query and navigate among those.

One tricky thing about queries is that we don't get a "true" view of the contents of the file, because any macros will have already been evaluated and we just end up with the list of targets underneath. But there's enough information in the returned query data (the `generator_name` and `generator_location` fields) that would let us reconstruct a symbol hierarchy—if a macro creates a target, then we show the macro at the top level and then expand it to show the actual targets underneath.

...alright, I can see a path forward now where this works really nicely. Let's go ahead and get this in and then iterate over the other enhancements I just talked about. Can you rebase your change so it has the new query stuff I merged this morning, and then I can take a closer look at the details?

File: src/symbols/bazel_build_symbol_provider.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+import {
+  getBazelWorkspaceFolder,
+  getTargetsForBuildFile,
+  QueryLocation,
+} from "../bazel";
+import { blaze_query } from "../protos";
+
+/** Provids Symbols for targets in Bazel BUILD files. */
+export class StarlarkSymbolProvider implements DocumentSymbolProvider {
```
Comment (allevato) [inline src/symbols/bazel_build_symbol_provider.ts]: Can you rename this to be `BazelTargetSymbolProvider` (and rename the file to `bazel_target_symbol_provider.ts`? That will more accurately reflect what kind of "symbols" we're providing here.

File: src/bazel/bazel_utils.ts [hunk 2]
```diff
@@ -15,6 +15,10 @@
 import * as fs from "fs";
 import * as path from "path";
 
+import { blaze_query } from "../protos";
```
[the remark this answers is not in the document]
Comment (damienpontifex) [reply src/bazel/bazel_utils.ts]: It was. Had trouble finding exactly where, but it was under a clang format setting also trying to apply for TypeScript. Hopefully resolved now

**Comment by allevato** [comment]
Thanks!

**Comment by Globegitter** [comment]
@damienpontifex I am just trying out the extension from latest master which includes this PR - I can see the symbols in a build file pressing `Ctrl + Shift + O` but is this also meant to add `Go to definition`? I.e. if I ctrl+click on a dependecy of a target, like `"//my/target:go_default_library",` should this then bring me to the build file that includes this definition? That I think would be very useful.

**Comment by damienpontifex** [comment]
@Globegitter I agree that would be helpful 😀 and it's definitely something I'd like to add. This PR just covers the symbol search and jump to position in that file for now. Hopefully we can build on it to add this capability

**Comment by Globegitter** [comment]
Ok cool just wanted to check what exactly this was adding - great addition though, thanks :)

**Pull request merged**